### ISSUE TITLE
monkey.patch_all and gunicorn with more than 1 worker

### ISSUE THREAD

**Issue by emamirazavi** [benoitc/gunicorn#1056]
I'm using Flask+Gunicorn+Gevent
I've called monkey.patch_all at the first of app passed to gunicorn command line because of using some co-routine tasks in my app. 
My worker class is gunicorn.workers.ggevent.GeventWorker.
When i configure gunicorn to use more than one worker (multiprocessing.cpu_count() \* 2 + 1 that in my laptop is equal to 9) everything is mixed up. My logger (python logging) logs 10 times! and some much more others conflicts in my code occurred.
MY GUNICORN CONFIGURAION

```
# to import server_hook
import os, sys
this_path = os.path.dirname(os.path.realpath(__file__))
sys.path.insert(0, os.path.realpath(this_path))
import server_hook
import multiprocessing
user = "myapp"
max_requests = 60
max_requests_jitter = 5
reload = True
group = "daemon"
host = "0.0.0.0"
bind = ['127.0.0.1:8000', 'unix:/run/gunicorn/socket']
workers = multiprocessing.cpu_count() * 2 + 1
threads = 10
worker_class = "gunicorn.workers.ggevent.GeventWorker"
proc_name = "myapp"
default_proc_name = "myapp"
environment = LANG = "en_US.UTF-8"
preload_app = True
loglevel = "DEBUG"
on_starting = server_hook.on_starting
```

MY SERVER_HOOK FILE (server_hook.on_starting)

```
import threading
from myapp.logger import logger

def on_starting(server):
    thread = threading.Thread(target=test)
    thread.daemon = True
    thread.start()

def test():
    while 1:
        logger.debug('i am test!')
        sleep(10)
```

MY LOG FILE

```
2015-06-25 20:32:33,579 INFO Starting webserver...
2015-06-25 20:32:33,595 DEBUG i am test!

2015-06-25 20:32:43,595 DEBUG i am test!
2015-06-25 20:32:43,596 DEBUG i am test!
2015-06-25 20:32:43,596 DEBUG i am test!
2015-06-25 20:32:43,596 DEBUG i am test!
2015-06-25 20:32:43,595 DEBUG i am test!
2015-06-25 20:32:43,596 DEBUG i am test!
2015-06-25 20:32:43,597 DEBUG i am test!
2015-06-25 20:32:43,597 DEBUG i am test!
2015-06-25 20:32:43,597 DEBUG i am test!
2015-06-25 20:32:43,595 DEBUG i am test!

2015-06-25 20:32:53,596 DEBUG i am test!
2015-06-25 20:32:53,597 DEBUG i am test!
2015-06-25 20:32:53,597 DEBUG i am test!
2015-06-25 20:32:53,597 DEBUG i am test!
2015-06-25 20:32:53,597 DEBUG i am test!
2015-06-25 20:32:53,597 DEBUG i am test!
2015-06-25 20:32:53,598 DEBUG i am test!
2015-06-25 20:32:53,598 DEBUG i am test!
2015-06-25 20:32:53,597 DEBUG i am test!
2015-06-25 20:32:53,599 DEBUG i am test!

```

For now the solution is to call patch_all with thread=False or set gunicorn configuration to use just one worker. In patch_all with thread=False solution

```
  File "/home/m.emami/projects/python/whatsapp/src/lib/myapp/cron.py", line 42, in cron_main
    rss_robot_cron()
  File "/home/m.emami/projects/python/whatsapp/src/lib/myapp/rss_robot.py", line 155, in cron
    feed_all()
  File "/home/m.emami/projects/python/whatsapp/src/lib/myapp/rss_robot.py", line 151, in feed_all
    feed_pool.join()
  File "/usr/lib64/python2.7/site-packages/gevent/pool.py", line 100, in join
    self._empty_event.wait(timeout=timeout)
  File "/usr/lib64/python2.7/site-packages/gevent/event.py", line 77, in wait
    result = self.hub.switch()
  File "/usr/lib64/python2.7/site-packages/gevent/hub.py", line 331, in switch
    return greenlet.switch(self)
error: cannot switch to a different thread
```

I can not use any thread being gevent safe. It means just one solution exists and it's use gunicorn with one worker!!!
Any idea?!

**Comment by tilgovi** [comment]
Where are you calling `patch_all()`? You shouldn't need to do so yourself. The gevent worker will call it for you just after the worker process forks and before initializing your application.

I suspect what is happening is that with `preload_app` your application is loaded before the `on_starting` hook is called. If your application is calling `patch_all` then maybe you're creating a greenlet instead of a real thread. Then, when gunicorn forks the worker processes you get 10 of these greenlets (one in each worker) since the greenlets will exist in the fork, unlike a traditional thread.

You might try these things:

- Don't monkey patch yourself. Let gunicorn do it when it forks.
- Create the thread in the configuration file at module level and then just start it in the `on_starting` hook. Doing it this way should ensure that you create a real thread.

**Comment by emamirazavi** [comment]
As i already told i want to use Gevent in my app when i use `preload_app = True` in configuration and you know better than me that Gunicorn just uses Gevent for it's forks. When i call patch_all for **my app** threading must be Gevent formatted unless i use `thread=False` in patch_all's argument.
If i don't call patch_all by myself and leave it for Gunicorn to do some check reads this:
```
import socket
def on_starting(server):
    logger.debug(socket.socket.__module__)
```

```
2015-06-26 03:10:27,158 INFO Starting webserver...
2015-06-26 03:10:27,174 DEBUG socket
```
if gevent patch_all had been called with Gunicorn logger should log `gevent.socket`

**Comment by emamirazavi** [comment]
if gevent patch_all had been called with Gunicorn logger should log `gevent.socket`
I want to use Gevent feature in server side of my app that is different to web workers. It means i have some robots and some *co-routine tasks* that regardless to Gunicorn workers should be done and looped. It's better you offer a good solution to me in where i should call `monkey.patch_all`

**Comment by jamadden** [comment]
For various reasons we also need to make sure that gevent has monkey patched everything sooner than gunicorn wants to do it. Our (preloaded) app uses `pyramid`, which loads things with a script called `pserve`. We wrap this script in our own script, monkey patch there, and then hand things over to pserve, which hands them over to gunicorn. You could probably use a similar technique of wrapping whatever script you normally use.

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
A script to help us ensure that we are loading and monkey-patching
the entire system before Pyramid loads: loading Pyramid's ``pserve``
loads many Pyramid modules, including :mod:`pyramid.traversal`, which
in turn loads :mod:`repoze.lru` and allocates real, non-recursive
thread locks. These are not compatible with gevent and eventually
lead to a hang if we re-enter a greenlet that wants to acquire one
of these locks while a previous greenlet already has it.
"""

from __future__ import print_function, unicode_literals, absolute_import, division
__docformat__ = "restructuredtext en"

# Note that we must not import *anything* before the patch
from nti.monkey import gevent_patch_on_import
gevent_patch_on_import.patch() # basically does gevent.monkey.patch_all()

import sys
from pkg_resources import load_entry_point, get_distribution

def main():
	sys.exit(
		load_entry_point('pyramid', 'console_scripts', 'pserve')()
	)
```

**Comment by tilgovi** [comment]
Gunicorn will call `patch_all` for you but it will do so *after* `on_starting` and *if you use preload* this is also after your app loads, I think. That will present problems if you have imports at that point already (like `threading` or `socket` in your examples). I was trying to suggest a way to use just `threading` to purposefully get the original stdlib `threading` because it seemed you wanted the behavior of threads where the thread does not fork.

If you are trying to run a single background thread rather than one per worker you should probably handle it as a separate process with whatever process supervision tools you use and not try to have Gunicorn start it. Either that or you will probably have to stop using preload.

To be honest, I'm not even sure how safe preload is with gevent since anything that gets imported when your application's main module is imported might load modules before they are monkey patched. You will have to experiment and be very careful with how you access things. It will help somewhat to be sure that you import modules and then refer to their functions rather than using `from foo import some_function`.

**Comment by tilgovi** [comment]
Maybe we should introduce a way to use the gevent worker without automatic monkey patching. Then users can control when and what is patched for themselves.

**Comment by emamirazavi** [comment]
I think Gunicorn is not Gevent safe at other layers than create workers layer in `init_process` method.
```
#!/usr/bin/python
# EASY-INSTALL-ENTRY-SCRIPT: 'gunicorn==19.3.0','console_scripts','gunicorn'
__requires__ = 'gunicorn==19.3.0'

from gevent import monkey
monkey.patch_all()

import sys
from pkg_resources import load_entry_point

if __name__ == '__main__':
    sys.exit(
        load_entry_point('gunicorn==19.3.0', 'console_scripts', 'gunicorn')()
    )
```
If we patch all before starting Gunicorn, we must assure that Gunicorn is Gevent safe. When i patched all before starting Gunicorn, signalling did not work and systemd could not reload the app correctly. For now, regretfully, i'm supposed to run another process to do tail of my program!!! I'm very sorry. But nowadays it's better, software like Gunicorn, can handle applications in themselves. This prevents applications from parting to several ugly depended executive files. Any happy idea?

**Comment by benoitc** [comment]
@emamirazavi this is a wrong assumption. Monkey patching systems calls is not safe not the contrary, and it shouldn't be done at all in the arbiter since it would break all the signalling an other stuffs. 

**Monkey patching systems calls (sockets, threads, signaling) in the arbiter by itself is actually very dangerous and is strongly discouraged and not supported right now**. The nature of python doesn't allows us to force the use of original system system call until we know it before. Not sure right now if it's even possible to make sure we are always using the original functions. This should be investigated.

To answer to your question, if you need to monkey patch threds (which have been removed due to the number of application errors it creates when enabled by default) then you can do it in a [post_worker_init](http://docs.gunicorn.org/en/19.3/settings.html#post-worker-init) hook. I agree with @tilgovi we could make it more simpler.

Btw why do you need to preload your application?

> note: protect the core against possible monkey patching should be investigated since indeed it would make it resistant to any usage. I will work on it asap, any idea is welcome :)

**Comment by benoitc** [comment]
@emamirazavi also since gunicorn is already monkey patching in the worker why not only monkey patching threads in your application?

Another thing is that monkey patching globally like this is not really needed. Why not monkey patching once your application is initialised or on first request?

**Comment by emamirazavi** [comment]
> Btw why do you need to preload your application?
@benoitc I do some concurrent tasks more than requesting and responding at web layer.

> since gunicorn is already monkey patching in the worker why not only monkey patching threads in your application?
@benoitc You know that i don't want do my tasks in the worker! In every place except worker, monkey patching is not called!
@benoitc Everything being occurred has be issued at the first of this subject. Repeated logs and unusuall behaviors.

@benoitc How can i use gevent benefits with monkey patching on first request?!

**Comment by emamirazavi** [comment]
@benoitc Can i have some live workers and lock them to prevent be killed along side of web server workers?! I think this can solve my problem.

**Comment by benoitc** [comment]
@emamirazavi looks at the ticket history to understand why we don't do any thread monkey patching. Lot of applications don't support it. 

Gevent loop will be used at the first request, we already take care about that. 

Monkey patching before or after a request for threads doesn't change anything for your application. For pre-initialisation steps I suggest to launch a new process like suggested @tilgovi or a simple thread. You can launch anything you want in the arbiter using the hooks before the fork.

One possible example is starting a pipe, and launching a process that will share it with a worker. Or anything else. In near future this will be available in gunicorn.

**Comment by benoitc** [comment]
@emamirazavi in short the design of gunicorn prevent by default to share anything between the worker. Which doesn't mean it is isn't possible. There are different methods to do it using redis, sharing a pipe, or a memory mapped file, ...

**Comment by benoitc** [comment]
Revisiting this issue. 

So to be clear, monkey patching in the arbiter is not supported and should be discouraged. Preload should in fact for some edge cases where you need to reduce the memory usage.

To close the issue I am thinking to the following actions:
- document  the preload_app setting to make it clear what it can be used for and what is discouraged
- document how to share objects and data between workers in the current design
- open a ticket to implement an API to allows message passing between workers

Anything else?

**Comment by emamirazavi** [comment]
If for now gunicorn's support and vision is the same, you items you
mentioned are good and perfect. We should document that monkey patching
does not work in preload_app. Many thanks Benoit.

**Comment by benoitc** [comment]
@emamirazavi will do the work this week. Thanks for the feedback!

**Comment by emamirazavi** [comment]
Where should I do the tasks?!

**Comment by benoitc** [comment]
@emamirazavi you can update the doc in the `Running Gunicorn` section. It should be the right pqce IMO.

**Comment by emamirazavi** [comment]
It's done, give me a feedback.

**Comment by Anishmourya** [comment]
i had same issue . this is solved by using app_before_request hook . so call  background task function on any hook that would call after application initialised  by gunicorn .

**Comment by benoitc** [comment]
no activity since awhile. closing feel free to create a new ticket if needed.
